lib/fine-types/src/Language/FineTypes/Export/OpenAPI/Value/ToJSON.hs

```diff
@@ -47,28 +49,41 @@ import qualified Language.FineTypes.Typ as Typ
 -- We need the 'Typ' of the 'Value' in order to add field names.
 --
 -- Note: We assume that the 'Value' has the given 'Typ'.
+reportTypeValue :: (Show t, Show v) => t -> v -> String
+reportTypeValue typ value =
+    TL.unpack $ pShow (("type:" :: Text, typ), ("value:" :: Text, value))
+
+typeCheckAssertion :: Typ -> Value -> t -> t
+typeCheckAssertion typ value t =
+    if hasTyp value typ
+        then t
+        else error $ "typeCheckAssertion: " <> reportTypeValue typ value
+
+impossibleCase :: (Show t, Show v) => t -> v -> String -> a
+impossibleCase typ value prefix =
+    error $ prefix <> ": " <> reportTypeValue typ value
+
 jsonFromValue :: Typ -> Value -> JS.Value
-jsonFromValue = go
+jsonFromValue t0 v0 = typeCheckAssertion t0 v0 $ go t0 v0
```

Comment:
```suggestion
jsonFromValue = go
```

I'm not against type-checking the assertion per se, but the function `hasTyp` incurs a run-time cost that I would like to avoid. Could you shift `typeCheckAssertion` to the usage site, e.g. to the tests instead?